src/attacks.rs

```diff
@@ -14,15 +22,214 @@ pub enum DepthReduceSet {
+//  replace that.
+#[derive(Debug)]
+pub enum AttackTarget {
+    Depth(f64),
+    Size(f64),
+}
+
+/// Range of targets to try (to find the optimum value) from `start`, increasing
+/// by `interval` until `end` is reached or surpassed.
+// FIXME: Using this instead of `std::ops::Range<f64>` because Rust correctly
+//  doesn't allow iterating over floating point values but there is probably
+//  an easier way than coding this from scratch.
+// FIXME: Assert range validity in the struct itself instead of on the caller
+//  (`attack`).
+#[derive(Debug)]
+pub struct TargetRange {
+    pub start: f64,
+    pub interval: f64,
+    pub end: f64,
+}
+
+#[derive(Debug)]
+pub struct AttackProfile {
+    pub runs: usize,
+    pub target: AttackTarget,
+    pub range: TargetRange,
+    pub attack: DepthReduceSet,
+}
+
+impl AttackProfile {
+    /// Build a default profile from an attack type that has only one run
+    /// in a range of a single value (to make it compatible with previous
+    /// uses of `attack`).
+    // FIXME: We shouldn't need the `graph_size` (or the graph for that
+    // matter), but this is accommodating previous uses of `attack` (which
+    // should be refactored entirely and this method removed or reworked).
+    pub fn from_attack(attack: DepthReduceSet, graph_size: usize) -> Self {
+        let graph_size = graph_size as f64;
+        let target = match attack {
+            DepthReduceSet::ValiantDepth(depth) => AttackTarget::Depth(depth as f64 / graph_size),
+            DepthReduceSet::ValiantSize(size) => AttackTarget::Size(size as f64 / graph_size),
+            DepthReduceSet::ValiantAB16(size) => AttackTarget::Size(size as f64 / graph_size),
+            DepthReduceSet::GreedyDepth(depth, _) => AttackTarget::Depth(depth as f64 / graph_size),
+            DepthReduceSet::GreedySize(size, _) => AttackTarget::Size(size as f64 / graph_size),
+        };
+        // FIXME: This code should absorb the `depth_reduce` and derived
+        // functions logic. The target discrimination depth/size should
+        // b independent of the attack type (valiant/greedy).
+
+        let range = {
+            let single_value = match target {
+                AttackTarget::Depth(depth) => depth,
+                AttackTarget::Size(size) => size,
+            };
+            // FIXME: Too verbose, there probably is a more concise way to do this.
+            TargetRange {
+                start: single_value,
+                end: single_value,
+                interval: 0.0,
+            }
+        };
+
+        AttackProfile {
+            runs: 1,
+            target,
+            range,
+            attack,
+        }
+    }
+}
+
+/// Results of an attack expressed in relation to the graph size.
+#[derive(Clone, Default, Debug)]
+pub struct SingleAttackResult {
+    depth: f64,
+    exclusion_size: f64,
+    // graph_size: usize,
+    // FIXME: Do we care to know the absolute number or just
+    // relative to the graph size?
```

Comment:
In the end the fraction depth/graphsize and set/graphsize is what we use to compare